<a href="https://colab.research.google.com/github/nusc2016/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Aaron_Huizenga_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [0]:
## While this is more in depth and absolutely mind boggling, watching previous
## lecture videos has helped me a little bit, but I will tell you that this has 
## me freaking out to the core. 

import pandas as pd 
import numpy as np
import scipy as stats
import seaborn as sns


/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [0]:
df = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data", header=None)
df.head()

0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15 16
0  republican  n  y  n  y  y  y  n  n  n  y  ?  y  y  y  n  y
1  republican  n  y  n  y  y  y  n  n  n  n  n  y  y  y  n  ?
2    democrat  ?  y  y  ?  y  y  n  n  n  n  y  n  y  y  n  n
3    democrat  n  y  y  n  ?  y  n  n  n  n  y  n  y  n  n  y
4    democrat  y  y  y  n  y  y  n  n  n  n  y  ?  y  y  y  y

In [46]:
## I need to relabel the headers so that in checking the shape I will have a 
## better feel for what I'm looking at

df = df.rename(columns={0: 'party', 
                        1: 'handicapped-infants', 
                        2: 'water-project', 
                        3: 'budget', 
                        4: 'physician-fee-freeze', 
                        5: 'el-salvador-aid', 
                        6: 'religious-groups', 
                        7: 'anti-satellite-test-ban', 
                        8: 'aid-to-contras', 
                        9: 'mx-missile', 
                        10: 'immigration', 
                        11: 'synfuels', 
                        12: 'education', 
                        13: 'right-to-sue', 
                        14: 'crime', 
                        15: 'duty-free', 
                        16: 'south-africa'})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
## What I will do now is replace y and n values with 0 and 1 so that it makes 
## more sense to the naked eye

df = df.replace(['n', 'y'], [0, 1])
print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   0             1  ...     1         0            1
1  republican                   0             1  ...     1         0            ?
2    democrat                   ?             1  ...     1         0            0
3    democrat                   0             1  ...     0         0            1
4    democrat                   1             1  ...     1         1            1

[5 rows x 17 columns]

In [0]:
## As you can see there are 435 rows and 17 columns
## What I need to do for personal reasons is replace "?" with NaN values

df = df.replace('?', np.NaN)
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
df.isna().sum()

party                        0
handicapped-infants         12
water-project               48
budget                      11
physician-fee-freeze        11
el-salvador-aid             15
religious-groups            11
anti-satellite-test-ban     14
aid-to-contras              15
mx-missile                  22
immigration                  7
synfuels                    21
education                   31
right-to-sue                25
crime                       17
duty-free                   28
south-africa               104
dtype: int64

In [0]:
## Using sample from the lecture 

rep = df[df['party'] == 'republican']
dem = df[df['party'] == 'democrat']


In [0]:
from scipy.stats import ttest_ind, ttest_ind_from_stats, ttest_rel

##Education Bill##


In [0]:
## I hope that I am understanding the assignment correctly. This is what I have 
## come up with.

print("Democrat Support: ", dem['education'].mean())
print("Republican Support: ", rep['education'].mean())

Democrat Support:  0.14457831325301204
Republican Support:  0.8709677419354839


In [0]:
ttest_ind(dem['education'], rep['education'], nan_policy='omit')

Ttest_indResult(statistic=-20.500685724563073, pvalue=1.8834203990450192e-64)

-Null hypothesis: The mean of Democrat support is equal to the mean of Republican support for this bill.

-Alternative hypothesis: The mean of Democrat support is not the same as the mean of Republican support for this bill.

-T-statistic: -20.5

-P-value: 1.88

-Convention: 95%

**Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans**

##MX-Missile Bill##

In [0]:
print("Democrat Support: ", dem['mx-missile'].mean())
print("Republican Support: ", rep['mx-missile'].mean())

Democrat Support:  0.7580645161290323
Republican Support:  0.11515151515151516


In [0]:
ttest_ind(dem['mx-missile'], rep['mx-missile'], nan_policy='omit')

Ttest_indResult(statistic=16.437503268542994, pvalue=5.03079265310811e-47)

-Null hypothesis: The mean of Democrat support is equal to the mean of Republican support for this bill.

-Alternative hypothesis: The mean of Democrat support is not the same as the mean of Republican support for this bill.

-T-statistic: 16.4375

-P-value: 5.03

-Convention: 95%

**Given the results of the above test, I would REJECT the null hypothesis that there is the same level of support for this bill among Democrats and Republicans**

##Water-Project##

In [0]:
print("Democrat Support: ", dem['water-project'].mean())
print("Republican Support: ", rep['water-project'].mean())

Democrat Support:  0.502092050209205
Republican Support:  0.5067567567567568


In [0]:
ttest_ind(dem['water-project'], rep['water-project'], nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

-Null hypothesis: The mean of Democrat support is equal to the mean of Republican support for this bill.

-Alternative hypothesis: The mean of Democrat support is not the same as the mean of Republican support for this bill.

-T-statistic: -0.089

-P-value: 0.93

-Convention: 95%

**Given the results of the above test, I would fail to reject the null hypothesis that there is the same level of support for this bill among Democrats and Republicans**

## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!

### Stretch Goal #1 

- $H_0: \mu_r = 0.5$
- $H_0: \mu_r \neq 0.5$

- Confidence Level: 95%

In [0]:
## I have never attempted a stretch goal so this should be interesting. There was
## a clue given during the lecture as to what the stretch goal would look like,
## so I'm not exactly sure what will happen. I tried using test_data = test.random
## but it didn't work, so I switched to random_data = random.random() and that 
## isn't working either, so I am not going to attempt the stretch goal.